### ISSUE TITLE
Can't overrideDerivation on systemd

### ISSUE THREAD

**Issue by nh2** [NixOS/nixpkgs#30353]
I'm trying to patch systemd, but already using an empty `overrideDerivation` fails on it:

```
{
  nixpkgs.config.packageOverrides = {
    systemd = pkgs.lib.overrideDerivation pkgs.systemd (oldAttrs: rec {
    });
  };
}
```

`error: attribute ‘udev’ missing, at /home/niklas/src/nixpkgs/pkgs/os-specific/linux/lvm2/default.nix:31:38`

On nixpkgs `dea58ce2`.

Why is that and how to do it right?

**Comment by FRidh** [comment]
You tried using `overrideAttrs` instead?

**Comment by nh2** [comment]
Yes, `pkgs.systemd.overrideAttrs` resulted in the same problem.

**Comment by FRidh** [comment]
In any case, does it not make sense that it fails. You basically say here `derivation { }`, i.e., an empty derivation. `udev` is part of `systemd` if I'm correct so during evaluation references to `udev` will complain.

**Comment by nh2** [comment]
@FRidh Not sure I understand, with other packages I can do

```
{
  nixpkgs.config.packageOverrides = {
    systemd = pkgs.lib.overrideDerivation pkgs.mypackage (oldAttrs: rec {
      patches = [ ... ];
    });
  };
}
```

and it works perfectly file, because usually those attrs are just added to your derivation, right? And using `{ }` should have the original one unchanged, correct?

**Comment by dezgeg** [comment]
The problem is in `all-packages.nix`:
````
11945   systemd = callPackage ../os-specific/linux/systemd {
11946     utillinux = utillinuxMinimal; # break the cyclic dependency
11947     gperf = gperf_3_0; # fix build until v233
11948   }
11949     // {
11950       udev.bin = systemd;     # ${systemd.udev.bin}/bin/udevadm
11951       udev.lib = libudev.out; # ${systemd.udev.lib}/lib/libudev.*
11952     };
````

The `// {` etc. part needs to be moved inside of `systemd/default.nix` in order to keep overrideDerivation / overrideAttrs working.

**Comment by FRidh** [comment]
@nh2 see https://github.com/NixOS/nixpkgs/pull/30872

**Comment by t184256** [comment]
Are there any workarounds save for writing overlays?

**Comment by nh2** [comment]
@FRidh nice

**Comment by vcunat** [comment]
Fixed on master by #30872.

**Comment by nh2** [comment]
Thank you!
